Circular Barn - Game Theory editorial (#6260)
cpinitiative/usaco-guide: solutions/silver/usaco-1255.mdx

---
id: usaco-1255
source: USACO Silver 2022 December
title: Circular Barn
author: John Zhou
---

[Official Analysis (C++, Python)](https://usaco.org/current/data/sol_prob2_silver_dec22.html)

## Explanation

Below, $a_i$ can also refer to the current number of cows in the $i$-th room, and $\mathbb P$ is the set of primes including $1$.

First, consider the case $N=1$, when both farmers remove cows from one room.

Note that it is possible to remove $1$, $2$, or $3$ but not any multipes of $4$ cows.
We should therefore pay close attention fo $a_1\bmod 4$:
A "first" player starting with $a_1\bmod 4\neq 0$ should take $a_1\bmod 4$ cows to force the "second" player into $a_1\bmod 4=0$, who will always give the first player an $a_1\bmod 4\neq 0$, since they cannot take any multiple of $4$ to maintain the $0$ remainder.
The first player can now repeat the process to force the other player into a smaller $a_1\bmod 4=0$, which will eventually be $0$, causing the first player to win.

Therefore, when the initial state is $a_1\bmod 4\neq 0$, Farmer John, being the first player, will win,

> but if $a_1\bmod 4=0$ initially, then Farmer John must produce a state of $a_1\bmod 4\neq 0$ for Farmer Nhoj, forcing Farmer Nhoj to win instead.

Reviewer: ```suggestion
First, consider the case $N=1$, when both farmers remove cows from one room.

Each farmer can remove either $1$, $2$, $3$, but not $4$ or any multiples of it. This motivates us to consider $a_1 \bmod 4$ in our game. The player starting with $a_1\bmod 4\neq 0$ should take $a_1\bmod 4$ cows to force the other player into $a_1\bmod 4=0$, who will always give the first player an $a_1\bmod 4\neq 0$, since they cannot take any multiple of $4$ to maintain the $0$ remainder.
The first player can now repeat the process until $a_1 = 0$, resulting in a win.

Therefore, when $a_1 \bmod 4 \neq 0$, Farmer John will win. Otherwise, Farmer Nhoj will win.
```

idt you need to define $a_i$ and $P$ cuz $a_i$ is already given in the statement and $P$ is only used once in a kind of jargon way

> Now, we consider the case when $N>1$.  Note that operations on each room are all independent, so we can consider the result of playing each room as if $N=1$, and term guaranteed loser of that room the "defender" and the guaranteed winner of that room the "attacker".

Reviewer: ```suggestion
Now, we consider the case $N>1$.  Since each room is independent, we treat the overall game as $N$ individual games. For each room, we define the attacker as the guaranteed winner and the defender as the guaranteed loser.
```

the maximum $p<a_i$ is desired to provide the largest speedup to ending the game.
To compute the largest $p<x$ such that $p\equiv x\mod 4$, first use a sieve to find all primes, and keep an array $t_i$ with $0\leq i<4$ for the most recent prime $t_i=p$ we found where where $p\bmod 4=i$ and assign $m_x=t_{x\bmod 4}$ to represent the requested result, all as we go.
On the other hand, the defender can block this speedup by relying on the fact that there are no primes $p>4$ that satisfies $p\bmod 4=2$ (so no speedups when $a_i\bmod 4=2$).
If the defender takes $2$ cows from their $a_i\bmod 4=2=0$ and hand back $a_i\bmod 4=2$ each time, there will be no way for the attacker to speed up the game any more.

Thsu, when $a_i\bmod4\in\{1,3\}$, the only speedup the attacker can apply removing $m_{a_i}$ initially, and in all other cases there are no speedups.
When there are no speedups, $T_i=\frac{a_i}2$; when there is a speedup, $T_i=1+\frac{a_i-m_{a_i}}2$, as one operation is used to remove $m_{a_i}$.

We first compute $T_i$ for each room, and then then find the argmin of $\left\lceil\frac{T_i}2\right\rceil$ (if tied, with smallest $i$), and output the winner of that room.

> (Instead of using our initial strategy to compute the winner of a room, we could also check the parity of $T_i$.)

Reviewer: - right now, the writing is lowk a wall of paragraphs, which is kinda hard to read
  - I think breaking up some of the run-on sentences and spacing it out / using bullet points or other devices when appropriate would help
- I think a lot of this could be explained in more concise/less jargony ways?

for example, here's how I'd rewrite the first highlighted paragraph:
> Since the game ends at the first room with no cows, we can define the goals for the attacker and defender of each room:
> - The **attacker** wants to minimize the number of moves needed for the room's game to end, with the goal of ending the game in this room.
> - The **defender** wants to maximize the number of moves for the room's game to end, with the goal of prolonging this game for one of their rooms to result in a win.
> 
> For a room $i$, let the number of turns under optimal play be $T_i$. The winner of the game is determined by the room with the smallest value of $\lceil \frac{T_i}{2} \rceil$, with ties chosen by choosing the smallest index.

Author: I think your example is still jargony... I made my own changes, which are hopefully easier to read.

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>
<CPPSection>

> ```cpp

Reviewer: can you add more new lines between parts of your code

rn it's kind of crammed together

This means that the only speedup the attacker can apply is when $a_i\bmod4\in\{1,3\}$ (which also implies he goes first), removing $m_{a_i}$ initially;
After that, or in other cases, either there are no speedups available, or the defender will block all of them.
When there are no speedups, there are obviously $T_i=\frac{a_i}2$ operations only (in which $T_i$ represents the number of turns at this room);
when there is a speedup, there are $T_i=1+\frac{a_i-m_{a_i}}2$, as one operations is used to remove $m_{a_i}$.

We first compute the amount of operations ($T_i$) for each room and determine the number of cows in it $\bmod 4$ to determine the winner for that room;
we then find the argmin of $\left\lceil\frac{T_i}2\right\rceil$ (if tied, with smallest $i$), and output the winner of that room.
In practice, however, it becomes unneccessary to track the room, as the parity of the number of operations already encodes the winner.

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>
<CPPSection>

```cpp
#include <cstring>
#include <iostream>
using namespace std;
constexpr int MAX_NUMBER = 5000000;

> int isprime[MAX_NUMBER + 1];

Reviewer 2: snakecase isprime for consistency

int m[MAX_NUMBER + 1];

int main() {
	int T;
	cin >> T;

> 	memset(isprime, true, sizeof isprime);

Reviewer 2: Could this just use vectors instead? I think `memset` is unnecessary when you could just do `vector<int> is_prime(MAX_NUMBER, true)` .

Author: I hate vectors... but yeah will do

	isprime[1] = false;
	// 1 is considered a prime under this problem, even though
	// we don't sieve it; so assign 1 initially.

> 	int t[] = {-1, 1, -1, -1};

Reviewer 2: Can there be better variable names than `t`, `m`, and `minn` (below).

	m[1] = 1;
	// Sieve algorithm
	for (int i = 2; i <= MAX_NUMBER; i++) {
		if (isprime[i]) {
			t[i % 4] = i;
			for (int j = i; j < MAX_NUMBER; j += i) isprime[j] = false;
		}
		m[i] = t[i % 4];
	}
	while (T--) {
		int n;
		cin >> n;
		int minn = 2e9;
		for (int i = 1; i <= n; i++) {
			int a;
			cin >> a;
			int min_turns;
			if (a % 2 == 0) {
				min_turns = a / 2;
			} else {
				min_turns = 1 + (a - m[a]) / 2;
			}

> 			if (min_turns / 2 < minn / 2) { minn = min_turns; }

Reviewer 2: ```suggestion
			if (min_turns / 2 < minn / 2) minn = min_turns;
```

Nitpick but if this is on the same line brackets are not needed.

Reviewer 2: Also I have not read the explanation yet (lot of words 😭) but unless you want to use integer division properties wouldn't it be simpler to just write `min(min_turns, minn)`

Author: > Also I have not read the explanation yet (lot of words 😭) but unless you want to use integer division properties wouldn't it be simpler to just write `min(min_turns, minn)`

Well integer division properties is the exact point... although taking the min of turns / 2 will probably be more explicit.  gonna refactor this whole thing myself.

		}

> 		cout << ((minn & 1) ? "Farmer John" : "Farmer Nhoj") << endl;

Reviewer 2: ```suggestion
		cout << ((minn % 2 != 0) ? "Farmer John" : "Farmer Nhoj") << endl;
```

`minn & 1` is unnecessary complication with no performance difference.

The first player can now repeat the process until $a_1 = 0$, resulting in a win.

Therefore, when $a_1 \bmod 4 \neq 0$, Farmer John will win. Otherwise, Farmer Nhoj will win.

Now, we consider the case $N>1$.  Since each room is independent, we treat the overall game as $N$ individual games. For each room, we define the attacker as the guaranteed winner and the defender as the guaranteed loser.

The next pair of moves on a room will only be done when all the other rooms have been visited.
Thus, the players use the following strategies:
- The attacker in a room reduces the number of cows as fast as possible to return to this room and potentially win sooner.
- The defender would like this to play slow instead to maximize the chances of winning early at another room.
Let the total amount of operations to finish room $i$ individually be $T_i$.

> The game's winner will be the winner of the room with the minimum $\left\lceil\frac{T_i}2\right\rceil$ (if tied, the one with the minimum room number).

Reviewer: you need to add a new line here, cuz mdx thinks this is part of the bullet point

Author: ```suggestion
Let the total amount of operations to finish room $i$ individually be $T_i$.
The game's winner will be the winner of the room with the minimum $\left\lceil\frac{T_i}2\right\rceil$ (if tied, the one with the minimum room number).

```

we want the largest $p<a_i$ to give the largest speedup.
To compute the largest $p<x$ such that $p\equiv x\mod 4$, first use a sieve to find all primes, and keep an array $t_i$ with $0\leq i<4$ for the most recent prime $t_i=p$ we found where where $p\bmod 4=i$ and assign $m_x=t_{x\bmod 4}$ to represent the requested result, all as we go.
On the other hand, the defender can block this speedup by relying on the fact that there are no primes $p>4$ that satisfies $p\bmod 4=2$ (so no speedups when $a_i\bmod 4=2$).
If the defender takes $2$ cows from their $a_i\bmod 4=2=0$ and hand back $a_i\bmod 4=2$ each time, there will be no way for the attacker to speed up the game any more.

Thus:
- When $a_i\bmod4\in\{1,3\}$, the only speedup the attacker can apply is removing $m_{a_i}$ initially, resulting in $T_i=1+\frac{a_i-m_{a_i}}2$.
- Otherwise, there are no speedups with $T_i=\frac{a_i}2$.

We first compute $T_i$ for each room, and then then find the argmin of $\left\lceil\frac{T_i}2\right\rceil$ (if tied, with smallest $i$), and output the winner of that room.

> (Instead of using our initial strategy to compute the winner of a room, we could also check the parity of $T_i$.)

Reviewer: ```suggestion
To minimize the number of rounds, the attacker can use a modified version of his strategy when $N=1$. Instead of taking away $a_i\bmod 4$ cows, they can take away any $p\equiv a_i\mod 4$ cows as long as $p$ is prime and less than $a_i$.
To compute the largest $p<x$ such that $p\equiv x\mod 4$, first use a sieve to find all primes, and keep an array $t_i$ with $0\leq i<4$ for the most recent prime $t_i=p$ we found where $p\bmod 4=i$. As we iterate on values of $x$, we store $m_x=t_{x\bmod 4}$ to represent the largest equivalent prime.

On the other hand, the defender can block this speedup by relying on the fact that there are no primes $p>4$ that satisfy $p\bmod 4=2$, resulting in no speedups when $a_i\bmod 4=2$.
If the defender takes $2$ cows from their $a_i\bmod 4=2=0$ and hand back $a_i\bmod 4=2$ each time, there will be no way for the attacker to speed up the game anymore.

Thus:
- When $a_i\bmod4\in\{1,3\}$, the only speedup the attacker can apply is removing $m_{a_i}$ initially, resulting in $T_i=1+\frac{a_i-m_{a_i}}2$.
- Otherwise, there are no speedups with $T_i=\frac{a_i}2$.

We first compute $T_i$ for each room, and then find the minimum of $\left\lceil\frac{T_i}2\right\rceil$ (if tied, with smallest $i$), and output the winner of that room.
```

- fixed some grammar stuff (grammarly is good for proofreading some basic stuff)
- just some syntax / tone cleanup

First, let's approach this problem when $N=1$, which is when when both farmers remove cows from one room until it is empty.

Notice that it is possible to remove $1$, $2$, or $3$ but not any multipes of $4$ cows.
This implies that we should try to center our strategy around manipulating $a_1\bmod 4$.
First, if a player arrives in a state where $a_1\bmod 4=0$, then they cannot get out of it, as there are no multiples of $4$.
On the other hand, if a player arrives in a state where $a_1\bmod 4\neq 0$, they could always take away $a_1\bmod 4\in\{1,2,3\}$ to force the other player into a state where $a_1\bmod 4=0$, which they cannot get out of.

Our strategy for $N=1$ thus falls out immediately:
A player with a state of $a_1\bmod 4\neq 0$ should always take away $a_1\bmod 4$ cows to force the other player into a state where $a_1\bmod 4=0$;
that player will always return with a state of $a_1\bmod 4\neq 0$, and you can repeat this and give them a smaller $a_1\bmod 4=0$, which will eventually choke them down into $0$, which is when you win.
Therefore, when the initial state is $a_1\bmod 4\neq 0$, Farmer John will win;

> but if $a_1\bmod 4=0$ initially, then Farmer John must produce a state of $a_1\bmod 4\neq 0$ for Farmer Nhoj, forcing Farmer Nhoj to win instead.

Reviewer: should condense this section

also I (think) most/many people found the pattern just thru trial and error? at least idt it's necessarily natural to observe 1, 2, 3 and figure out mod 4, esp for a silver contestant

Author: Done

After that, or in other cases, either there are no speedups available, or the defender will block all of them.
When there are no speedups, there are obviously $T_i=\frac{a_i}2$ operations only (in which $T_i$ represents the number of turns at this room);
when there is a speedup, there are $T_i=1+\frac{a_i-m_{a_i}}2$, as one operations is used to remove $m_{a_i}$.

We first compute the amount of operations ($T_i$) for each room and determine the number of cows in it $\bmod 4$ to determine the winner for that room;
we then find the argmin of $\left\lceil\frac{T_i}2\right\rceil$ (if tied, with smallest $i$), and output the winner of that room.
In practice, however, it becomes neccessary to only track the number of operations itself, as the parity of the number of operations already encodes the winner.

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>
<CPPSection>

```cpp
#include <cstring>
#include <iostream>
#include <vector>
using namespace std;
constexpr int MAX_NUMBER = 5e6;
constexpr int MAX_INT = 2e9;

int main() {
	vector<bool> is_prime(MAX_NUMBER + 1, true);
	// This is m[] in the explanation.

> 	vector<int> largest_prime_equiv(MAX_NUMBER + 1);

Reviewer: descriptive variable names are good but IMO these are kind of long

rn my brain is too fried to come up w/ better alternatives but it'd be good if you could find smth a bit more concise

Author: I can’t come up with anything either.  If anything lmk

	is_prime[1] = false;
	// 1 is considered a prime under this problem, even though
	// we don't sieve it; so assign 1 initially.
	// This is t[] in the explanation.
	int largest_prime_mod[] = {-1, 1, -1, -1};
	largest_prime_equiv[1] = 1;
	// Sieve algorithm
	for (int i = 2; i <= MAX_NUMBER; i++) {
		if (is_prime[i]) {
			largest_prime_mod[i % 4] = i;
			for (int j = i; j < MAX_NUMBER; j += i) is_prime[j] = false;
		}
		largest_prime_equiv[i] = largest_prime_mod[i % 4];
	}

	int test_cases;
	cin >> test_cases;
	for (int t = 0; t < test_cases; t++) {
		int n;
		cin >> n;
		// Minimum of turns, i.e. operations / 2
		int min_turn = MAX_INT;
		// Number of operations on the minimum-turns room.
		int min_turn_operations = MAX_INT;
		for (int i = 1; i <= n; i++) {
			int cows;
			cin >> cows;
			int operations;
			if (cows % 2 == 0) {
				operations = cows / 2;
			} else {
				operations = 1 + (cows - largest_prime_equiv[cows]) / 2;
			}
			if (operations / 2 < min_turn) {
				min_turn = operations / 2;
				min_turn_operations = operations;
			}
		}
		cout << ((min_turn_operations % 2 != 0) ? "Farmer John" : "Farmer Nhoj")

> 		     << endl;

Reviewer: formatting is weird

i think if the var was called like `min_ops` or smth then it'd be fine

Author: Well min_turn_operations is the number of operations for the room with the minimum turns... `min_turn_ops` perhaps?  Doing that.
